### PR TITLE
Website global search feature

### LINKED ISSUE #18234: [Website] Global Search Feature
##  Description
MXNet website hosts not only rich and comprehensive documentations, but also many useful resources including tutorials, installation guide, APIs and blogs etc. The majority of our target users are those who are curious about MXNet and want to explore, their primary target will mostly be tutorials, FAQs, blog posts and other solution contents. But the current version of MXNet website does not provide anything for them to find the materials they need. This could prevent users from further exploring and staying interested in MXNet. And for power users who will come to API docs directly, it will greatly improve their experience if a global search feature is available everywhere not only within API document.

## Proposed Solution
Global search feature on MXNet website can improve user experience, website usability and better convey our product values to end users.  A search bar should be implemented on the top Nav bar accessible to users across all main information pages, with the capability of searching information across all site and always provide the valid and fresh results.
Integrate a stable and trust worthy 3rd party search service to MXNet websiteIt is a lot more cost and time efficient than start from scratch. [DocSearch by Algolia](https://docsearch.algolia.com/) is one of the site search solutions that is technical documentation oriented and widely adopted by many professional documentation websites including [PyTorch](https://pytorch.org/), [TypeScript](https://www.typescriptlang.org/index.html ), [React](https://reactjs.org/), [Bootstrap](https://getbootstrap.com/), [Scala](https://www.scala-lang.org/) and [GitLab](https://about.gitlab.com/) etc. 

## Deliverable
A search bar powered by DocSearch on MXNet website, capable of searching across all site in any version, located on the right side of main Nav bar with a drop down menu attached that contains all available version options.

## Acceptance Criteria
* Search bar accessible across all information web pages
* Providing real time searching, search as you type, always return fresh and valid result
* The scope of search should be the whole website and all versions
* Simple and consistent front end user interface for searching and result displaying

## Stakeholders
@sandeep-krishnamurthy 
@aaronmarkham 
@leezu 
@sojiadeshina 
@szha 

## Reference
1. [Detailed quip project proposal - internal](https://quip-amazon.com/3bsRAMn6uvx0/MXNet-Website-20-Global-Search)
2. [DocSearch website](https://docsearch.algolia.com/)
3. [Please share your feedback here](https://quip-amazon.com/awmoAhi6FQuK/Website-20-Global-Search-Feedback)

## Dev link
Demo link of this project:  http://ec2-34-219-134-42.us-west-2.compute.amazonaws.com/

### PR DESCRIPTION
**Pull request by ys2843**
## Description ##
This PR implements the [website global search feature outlined in this GitHub issue](https://github.com/apache/incubator-mxnet/issues/18234). It added a fully functional search bar powered by DocSearch, capable of searching across all site in any version, located on the right side of main Nav bar with a drop down menu attached that contains all available version options.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
M `docs/static_site/src/_includes/head.html`
M `docs/static_site/src/_includes/header.html`
M ` docs/static_site/src/assets/main.scss`
M `docs/static_site/src/_sass/minima/_layout.scss`
Add `docs/static_site/src/assets/img/search-icon.svg`
Add `docs/static_site/src/assets/js/globalSearch.js`
Add `docs/static_site/src/_sass/globalSearch.scss`

## Comments ##
- Preview link: http://ec2-user@ec2-34-219-134-42.us-west-2.compute.amazonaws.com/
- Please note the preview only has global search on master version, in production 1.6 will also be supported 
- This PR is ready to merge

### CODE REVIEW

**Comment by ys2843** [comment]
@mxnet-label-bot add pr-work-in-progress

File: docs/static_site/src/_includes/header.html [hunk 1]
```diff
@@ -35,8 +35,57 @@
...
+          <input id="global-search" type="text" title="Search" placeholder="Search" />
+          <div id="global-search-dropdown-container">
+            <button class="gs-current-version btn" type="button" data-toggle="dropdown">
+                <span id="gs-current-version-label">1.6</span>
+                <svg class="gs-dropdown-caret" viewBox="0 0 32 32" class="icon icon-caret-bottom" aria-hidden="true">
+                    <path class="dropdown-caret-path" d="M24 11.305l-7.997 11.39L8 11.305z"></path>
+                </svg>
+            </button>
+            <ul class="gs-opt-group gs-version-dropdown">
+                <li class="gs-opt gs-versions">master</li>
```
Comment (marcoabreu) [inline docs/static_site/src/_includes/header.html:50]: can we generate that list?

Comment (ys2843) [reply docs/static_site/src/_includes/header.html:50]: Good suggestion, I will put logic for generating list in JS

**Comment by ChaiBapchya** [comment]
Search feature looks great.
Nice to see we leveraging third-party to solve our search problem.
Q1. This feature is currently only for landing page right?
It isn't available on other pages right?
For e.g. not available in Python API > NDArray page right?
http://ec2-34-219-134-42.us-west-2.compute.amazonaws.com/api/python/docs/api/ndarray/index.html

Can we document this point. And also add this functionality to rest of the pages in upcoming PRs?

Q2
When I search for copyto in 1.6 version on a laptop browser
It navigates to  https://mxnet.apache.org/versions/1.6/api/python/docs/api/ndarray/ndarray.html?astype=#ndarray

Isn't it supposed to go to the copyto operator specifically instead of the head of the div [ndarray]

**Comment by ys2843** [comment]
> Search feature looks great.
> Nice to see we leveraging third-party to solve our search problem.
> Q1. This feature is currently only for landing page right?
> It isn't available on other pages right?
> For e.g. not available in Python API > NDArray page right?
> http://ec2-34-219-134-42.us-west-2.compute.amazonaws.com/api/python/docs/api/ndarray/index.html
> 
> Can we document this point. And also add this functionality to rest of the pages in upcoming PRs?
> 
> Q2 Is search suggestion pointing to the correct place?
> When I search for copyto in 1.6 version on a laptop browser
> It navigates to https://mxnet.apache.org/versions/1.6/api/python/docs/api/ndarray/ndarray.html?astype=#ndarray
> 
> Isn't it supposed to go to the copyto operator specifically instead of the head of the div [ndarray]

Thank you for reviewing! Good questions.
Q1: Yes, it is available on all landing pages (that don't have a search bar currently), and not on any of the API docs pages. The reason is that while there is already a search bar on all API docs pages, in my opinion adding another one can confuse users. [Here is an example](https://pytorch.org/docs/stable/index.html), I think 1 search bar per page looks clear.
Q2: Good point. This is related to improvement of search result, I am currently working on these issues. It is mainly about how we configure the web crawler to grab data from our website, and it does not require any code changes to MXNet website.

File: docs/static_site/src/_includes/head.html [hunk 2]
```diff
@@ -5,14 +5,15 @@
...
+  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
   <link rel="stylesheet" href="{{ "/assets/main.css" | relative_url }}">
   {%- feed_meta -%}
   {%- if jekyll.environment == 'production' and site.google_analytics -%}
     {%- include google-analytics.html -%}
   {%- endif -%}
-
   <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
-
+  <script src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"></script>
```
Comment (aaronmarkham) [inline docs/static_site/src/_includes/head.html:15]: If we port a version that's hosting at a CDN in China, won't we need the files to be local?

File: docs/static_site/src/assets/js/globalSearch.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+ */
+
+/* Installation page display functions for install selector.
+   This utility allows direct links to specific install instructions.
+*/
+
+$(document).ready(function () {
+  const DEFAULT_CURRENT_VERSION = "1.6";
+  const VERSIONS = [
+    "master",
```
Comment (aaronmarkham) [inline docs/static_site/src/assets/js/globalSearch.js:27]: Will need to update the [release manager runbook](https://cwiki.apache.org/confluence/display/MXNET/Release+Process), so they know that this needs updating... like the 1.7 release going on right now...

File: docs/static_site/src/_includes/head.html (hunk 2 above)
Comment (ys2843) [reply docs/static_site/src/_includes/head.html:15]: Good question, I have just researched this CDN `www.jsdelivr.com` and based on its description about its [network](https://www.jsdelivr.com/network): 
> Big network in China
jsDelivr is the only public CDN with a valid ICP license issued by the Chinese government, and hundreds of locations directly in Mainland China.

I think it should work just fine, might load faster than hosting locally.

**Comment by ys2843** [comment]
> Looks good.
> 
> * Should we use offsite js files or host locally?
> * Should the API key be in a js file are can it be in a config file so we keep this stuff in someplace central?
> * Wonder if we could have the list of versions also in a config file and not in .js, so it is more obvious and easier to maintain... then other parts of the site like the install selector or other things that need the list of versions can share?
> 
> Just some thoughts... otherwise this looks good to me.
> Good work!

Should we use offsite js files or host locally?
+ Based on the research, the CDN works pretty well in China. I will keep an eye on it, if there is problem loading this file in some country, I will switch to host the file locally.

Should the API key be in a js file are can it be in a config file so we keep this stuff in someplace central?
+ Thanks for the suggestion. I will move API key to Jekyll config file

Wonder if we could have the list of versions also in a config file and not in .js, so it is more obvious and easier to maintain... then other parts of the site like the install selector or other things that need the list of versions can share?
+ Great idea, let me remove it from JS and set these common variables as Jekyll global var

File: docs/static_site/src/assets/js/globalSearch.js (hunk 3 above)
Comment (ys2843) [reply docs/static_site/src/assets/js/globalSearch.js:27]: Got it, I will move these to Jekyll config and update the runbook with the new place

Comment (sandeep-krishnamurthy) [reply docs/static_site/src/assets/js/globalSearch.js]: +1

**Comment by sandeep-krishnamurthy** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by sandeep-krishnamurthy** [comment]
@szha / @leezu - FYI.

**Comment by ys2843** [comment]
> Very nice. Thank you @ys2843 This will be very helpful for the community.
> Please update the release manager checklist for changing the defaults.
> 
> Should we change anything with algolia (like default version) whenever there is a new release?
> Also, for updating the search results, what is the process? What do you mean by improving the search result.
> 
> Overall, this is really nice improvement for easily finding the content. Thank you.

Thank you for reviewing. Will update the release manager checklist.
> Should we change anything with algolia (like default version) whenever there is a new release?

No, we don't need any change from algolia. We only need to update website code to set the default version. I will document this part.

> Also, for updating the search results, what is the process? What do you mean by improving the search result.

To make the search result more relevant and accurate, the process is to update the [configuration of DocSearch's web crawler](https://github.com/algolia/docsearch-configs/blob/master/configs/apache_mxnet.json). 
Currently the config is not ideal, and the best way to update and test the config is to run the web crawler locally. I have set up the crawler on an ec2 and run it with new config, and then make adjustment according to the result from the crawler.

**Comment by ys2843** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by ys2843** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by ys2843** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by ys2843** [comment]
@mxnet-label-bot add [website]

File: docs/static_site/src/_config.yml [hunk 4]
```diff
@@ -37,6 +37,18 @@ description: >- # this means to ignore newlines until "baseurl:"
 twitter_username: apachemxnet
 github_username:  apache/incubator-mxnet
 youtube_username: apachemxnet
+default_version: 1.6
```
Comment (marcoabreu) [inline docs/static_site/src/_config.yml:40]: Don't we have the versions as input parameter for the Jenkins job? Can't we just use those values? The fewer duplications, the better.

Comment (ys2843) [reply docs/static_site/src/_config.yml]: Removed this `default_version` key, while it is only used for version drop down default selection and it could always be the newest version below master.

**Comment by ys2843** [comment]
> Thank you!
> 
> Is it intended that the search queries entered while on the master version of the website yield results on 1.6 branch? I think it's better to be consistent
> 
> What's the plan to add the search button on other parts of the page besides the index page?

Thanks for the feedback!
Is it intended that the search queries entered while on the master version of the website yield results on 1.6 branch? I think it's better to be consistent
+ The search UI is consistent across all pages. There is a version select drop down on the right side of the search input field. It is currently set to 1.6 by default and the search results are only relevant to that version.

What's the plan to add the search button on other parts of the page besides the index page?
+ This feature is added to all information pages (tutorials, blogs, ecosystem, FAQ etc) on version master and 1.6 website. Not planned to add it to API pages while there is already a search bar on those pages.

**Comment by leezu** [comment]
Should the version dropdown been made consistent with the global version dropdown? Or at least default to the same version number as used in the global version dropdown?

> Not planned to add it to API pages while there is already a search bar on those pages.

Should we replace that search bar? It seems your search implementation is superior?

**Comment by ys2843** [comment]
> Should the version dropdown been made consistent with the global version dropdown? I mean at least default to the same version number as used in the global version dropdown?
> 
> > Not planned to add it to API pages while there is already a search bar on those pages.
> 
> Should we replace that search bar? It seems your search implementation is superior?

Good point, I will make the default version the same as the dropdown version.
There is plan to improve the search bar in API sites, but that wasn't discussed in this project. For this project the objective is to add search for main site.

**Comment by ys2843** [comment]
@mxnet-bot run ci [unix-cpu, website]
